atc/worker/fetcher.go

```diff
@@ -0,0 +1,170 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"io"
+	"time"
+
+	"github.com/concourse/concourse/atc/runtime"
+
+	"github.com/concourse/concourse/atc/resource"
+
+	"code.cloudfoundry.org/clock"
+	"code.cloudfoundry.org/lager"
+	"github.com/concourse/concourse/atc/db"
+	"github.com/concourse/concourse/atc/db/lock"
+)
+
+const GetResourceLockInterval = 5 * time.Second
+
+var ErrFailedToGetLock = errors.New("failed to get lock")
+var ErrInterrupted = errors.New("interrupted")
+
+//go:generate counterfeiter . Fetcher
+
+type Fetcher interface {
+	Fetch(
+		ctx context.Context,
+		logger lager.Logger,
+		containerMetadata db.ContainerMetadata,
+		gardenWorker Worker,
+		containerSpec ContainerSpec,
+		processSpec runtime.ProcessSpec,
+		resource resource.Resource,
+		owner db.ContainerOwner,
+		imageFetcherSpec ImageFetcherSpec,
+		cache db.UsedResourceCache,
+		lockName string,
+	) (GetResult, Volume, error)
+}
+
+func NewFetcher(
+	clock clock.Clock,
+	lockFactory lock.LockFactory,
+	fetchSourceFactory FetchSourceFactory,
+) Fetcher {
+	return &fetcher{
+		clock:              clock,
+		lockFactory:        lockFactory,
+		fetchSourceFactory: fetchSourceFactory,
+	}
+}
+
+type fetcher struct {
+	clock              clock.Clock
+	lockFactory        lock.LockFactory
+	fetchSourceFactory FetchSourceFactory
+}
+
+func (f *fetcher) Fetch(
+	ctx context.Context,
+	logger lager.Logger,
+	containerMetadata db.ContainerMetadata,
+	gardenWorker Worker,
+	containerSpec ContainerSpec,
+	processSpec runtime.ProcessSpec,
+	resource resource.Resource,
+	owner db.ContainerOwner,
+	imageFetcherSpec ImageFetcherSpec,
+	cache db.UsedResourceCache,
+	lockName string,
+) (GetResult, Volume, error) {
+	result := GetResult{}
```

Comment:
if I understood correctly, we don't really need this line - `result` is already being created at line 98